### DISCUSSION TITLE
Reflection: Easier reflection for known members

### DISCUSSION THREAD

**Discussion by FelixKlakow** [dotnet/csharplang#4627] [Language Ideas] [answered]
There are still no better way in C# for doing reflection for known Public members than the usual methods found among the API.
Sometimes this is pain to work with as it's unclear during compile time if you maybe got none or the wrong member. I suggest that we should be able to do things like this:

 ```csharp
static readonly MethodInfo MyMethod = reflect(Enumerable.Average(IEnumerable<decimal>));
static readonly MethodInfo MyMethod = reflect(Enumerable.Average(IEnumerable<object>)); // Compile time error as it doesn't exist
static readonly PropertyInfo MyProperty = reflect(MyType.MyProperty);
```

**Comment by YairHalberstadt** [comment, marked as answer]
This has been discussed previously at https://github.com/dotnet/csharplang/discussions/712

See also this blog post about why it's never gained much traction: https://docs.microsoft.com/en-gb/archive/blogs/ericlippert/in-foof-we-trust-a-dialogue

**Reply by FelixKlakow** [reply]
Sad, this is still definitely a feature a lot of people want when working a lot with reflection.

**Reply by YairHalberstadt** [reply]
Definitely, but it's never quite happened.

I imagine that if somebody were to do all the legwork of speccing and implementing it, then it would have a decent chance of being accepted, but that's a significant amount of work.

**Comment by HaloFour** [comment]
This is something that can be provided by tooling.  For example an analyzer could inspect the use of reflection APIs and compare it to the types being reflected and provide warnings.  IntelliJ does this with Java, for instance:

![image](https://user-images.githubusercontent.com/10102290/113864946-839eee00-9779-11eb-81e0-e12de008710a.png)
